src/server/v2.0/handler/project.go

```diff
@@ -2,33 +2,211 @@ package handler
+	"github.com/goharbor/harbor/src/lib/orm"
+	"github.com/goharbor/harbor/src/lib/q"
 	"github.com/goharbor/harbor/src/pkg/audit"
+	"github.com/goharbor/harbor/src/pkg/project/metadata"
+	"github.com/goharbor/harbor/src/pkg/quota/types"
+	"github.com/goharbor/harbor/src/pkg/retention/policy"
+	"github.com/goharbor/harbor/src/pkg/user"
+	"github.com/goharbor/harbor/src/replication"
+	"github.com/goharbor/harbor/src/server/v2.0/handler/model"
 	"github.com/goharbor/harbor/src/server/v2.0/models"
 	operation "github.com/goharbor/harbor/src/server/v2.0/restapi/operations/project"
 )
 
+const defaultDaysToRetention = 7
+
 func newProjectAPI() *projectAPI {
 	return &projectAPI{
-		auditMgr: audit.Mgr,
-		proCtl:   project.Ctl,
+		auditMgr:      audit.Mgr,
+		metadataMgr:   metadata.Mgr,
+		userMgr:       user.Mgr,
+		repositoryCtl: repository.Ctl,
+		projectCtl:    project.Ctl,
+		quotaCtl:      quota.Ctl,
 	}
 }
 
 type projectAPI struct {
 	BaseAPI
-	auditMgr audit.Manager
-	proCtl   project.Controller
+	auditMgr      audit.Manager
+	metadataMgr   metadata.Manager
+	userMgr       user.Manager
+	repositoryCtl repository.Controller
+	projectCtl    project.Controller
+	quotaCtl      quota.Controller
+}
+
+func (a *projectAPI) CreateProject(ctx context.Context, params operation.CreateProjectParams) middleware.Responder {
+	if err := a.RequireAuthenticated(ctx); err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	onlyAdmin, err := config.OnlyAdminCreateProject()
+	if err != nil {
+		log.Errorf("failed to determine whether only admin can create projects: %v", err)
+		return a.SendError(ctx, fmt.Errorf("failed to determine whether only admin can create projects: %v", err))
+	}
+
+	secCtx, _ := security.FromContext(ctx)
+	if onlyAdmin && !(secCtx.IsSysAdmin() || secCtx.IsSolutionUser()) {
+		log.Errorf("Only sys admin can create project")
+		return a.SendError(ctx, errors.ForbiddenError(nil).WithMessage("Only system admin can create project"))
+	}
+
+	req := params.Project
+
+	if req.RegistryID != nil && !secCtx.IsSysAdmin() {
+		// only system admin can create the proxy cache project
+		return a.SendError(ctx, errors.ForbiddenError(nil).WithMessage("Only system admin can create proxy cache project"))
+	}
+
+	// populate storage limit
+	if config.QuotaPerProjectEnable() {
+		if req.StorageLimit == nil || *req.StorageLimit == 0 || !secCtx.IsSysAdmin() {
+			setting, err := config.QuotaSetting()
+			if err != nil {
+				log.Errorf("failed to get quota setting: %v", err)
+				return a.SendError(ctx, fmt.Errorf("failed to get quota setting: %v", err))
+			}
+			defaultStorageLimit := setting.StoragePerProject
+			req.StorageLimit = &defaultStorageLimit
+		}
+	} else {
+		// ignore storage limit when quota per project disabled
+		req.StorageLimit = nil
+	}
+
+	if req.Metadata == nil {
+		req.Metadata = &models.ProjectMetadata{}
+	}
+
+	// accept the "public" property to make replication work well with old versions(<=1.2.0)
```

Comment:
do we still need it?

Reply:
Yes, we need this for the following will set the public metadata in it.

Reply:
Do you mean the following lines still need?

```
	if req.Public != nil && req.Metadata.Public == "" {
		req.Metadata.Public = strconv.FormatBool(*req.Public)
	}
```

I don't know whether it still needed, let's keep it here for the compatibility.